scipy/spatial/hausdorff_util.h

```diff
@@ -0,0 +1,59 @@
+// Utility C code for directed_hausdorff
+// Try to improve on the performance of the original Cython directed_hausdorff
+
+#include <math.h>
+
+struct return_values 
+{
+    double cmax;
+    int    index_1;
+    int    index_2;
+};
+
+// function for inner loop of directed_hausdorff algorithm
+void hausdorff_loop(const int data_dims,
+                    const double ar1[],
+                    const double ar2[],
+                    const int N1,
+                    const int N2,
+                    struct return_values * ret_vals)
+{
+    double               d, cmin, diff;
+    const double * const ar1_start_Ptr = ar1;
+    const double * const ar2_start_Ptr = ar2;
+    const int size_ar1 = data_dims * N1;
+    const int size_ar2 = data_dims * N2;
+    const double * const ar1_end_Ptr = &ar1[size_ar1 - 1];
+    const double * const ar2_end_Ptr = &ar2[size_ar2 - 1];
```

Comment:
This is a lot weirder than the typical index-based iteration style, and I'm fairly sure it compiles down to the same assembly either way.

In fact, I checked using a stripped-down version (including a call to `rand()` to make sure the loop doesn't get optimized away) and found that the normal syntax is the same or maybe better: https://godbolt.org/g/XBVNSa

Reply:
This is normally about as good. In fact you can check the code that cython produces. It's quite quite similar to this style.  In the end it's just written out what the compiler does when you use `ar[index]`.